### ISSUE TITLE
first update_state performed on keras.metric.Mean takes a long time for some models

### ISSUE THREAD

**Issue by geometrikal** [keras-team/keras#15831]
**System information**.
- Have I written custom code (as opposed to using a stock example script provided in Keras): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 20.04
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.6.0
- Python version: 3.8.10
- Bazel version (if compiling from source): N/A
- GPU model and memory: RTX 2080
- Exact command to reproduce: see MWE below

**Describe the problem**.

I have a couple of `keras.metrics.Mean` instances to track various loss values in my training step. When `update_state` is called on the first metric it takes a long time, but not on subsequent calls to other metrics. This happens even if I pass in a dummy value to update state.

Profiling shows that the slow down is happening here: https://github.com/keras-team/keras/blob/2c48a3b38b6b6139be2da501982fd2f61d7d48fe/keras/metrics.py#L478

Weirdly, the amount of time it takes is related to the size of the model

**Describe the current behavior**.

First call to `update_state` takes much longer than subsequent calls

**Describe the expected behavior**.

Call to `update_state` should take same amount of time

**Standalone code to reproduce the issue**.

```
import time
import tensorflow as tf
from keras import Input, Model
from keras.layers import Conv2D


def Backbone(filters):
    x = Input((None, None, 3))
    inp = x
    x = Conv2D(filters=filters, kernel_size=(3, 3), activation='relu')(x)
    x = Conv2D(filters=filters, kernel_size=(3, 3), activation='relu')(x)
    x = Conv2D(filters=filters*2, kernel_size=(3, 3), activation='relu')(x)
    x = Conv2D(filters=filters*2, kernel_size=(3, 3), activation='relu')(x)
    x = Conv2D(filters=filters*4, kernel_size=(3, 3), activation='relu')(x)
    x = Conv2D(filters=filters*4, kernel_size=(3, 3), activation='relu')(x)
    return Model(inp, x)


class TrainModel(tf.keras.Model):
    def __init__(self, *args, **kwargs):
        super(TrainModel, self).__init__(*args, **kwargs)
        self.loss_1 = tf.keras.metrics.Mean(name="loss_1")
        self.loss_2 = tf.keras.metrics.Mean(name="loss_2")
        self.loss_3 = tf.keras.metrics.Mean(name="loss_3")

    def train_step(self, data):
        with tf.GradientTape() as tape:
            res = self(data)
            loss_1 = tf.reduce_mean(res)
            loss_2 = tf.reduce_mean(res)
            loss_3 = tf.reduce_mean(res)
            loss = loss_1 + loss_2 + loss_3

        trainable_vars = self.trainable_variables
        gradients = tape.gradient(loss, trainable_vars)
        self.optimizer.apply_gradients(zip(gradients, trainable_vars))

        t0 = time.time()
        self.loss_1.update_state(1.0)
        t1 = time.time()
        self.loss_2.update_state(1.0)
        t2 = time.time()
        self.loss_3.update_state(1.0)
        t3 = time.time()
        print(f"{t1 - t0:.05f}s {t2 - t1:.05f}s {t3 - t2:.05f}s")

        return {m.name: m.result() for m in self.metrics}


print("filters = 8")
model = Backbone(filters=8)
model = TrainModel(model.inputs, model.outputs)
model.compile(optimizer='adam', run_eagerly=True)
model.fit(x=tf.random.normal([16,224,224,3]),
          batch_size=4,
          epochs=1,
          verbose=0)


print("filters = 32")
model = Backbone(filters=32)
model = TrainModel(model.inputs, model.outputs)
model.compile(optimizer='adam', run_eagerly=True)
model.fit(x=tf.random.normal([16,224,224,3]),
          batch_size=4,
          epochs=1,
          verbose=0)
```

Output is (note time to update is very slow):

```
filters = 8
0.00110s 0.00047s 0.00044s
0.00420s 0.00050s 0.00050s
0.00443s 0.00049s 0.00048s
0.00441s 0.00049s 0.00048s
filters = 32
0.00052s 0.00045s 0.00044s
0.04262s 0.00055s 0.00053s
0.04291s 0.00101s 0.00102s
0.03007s 0.00173s 0.00173s
```

**Work around**

Use custom metric:

```
import tensorflow as tf
from keras.metrics import Metric


class SimpleMean(Metric):
    def __init__(self, name="name", dtype=None):
        super(Metric, self).__init__(name=name, dtype=dtype)
        self.total = self.add_weight(
            'total', initializer='zeros')
        self.count = self.add_weight(
            'count', initializer='zeros')

    def update_state(self, value):
        try:
            value = tf.cast(value, self._dtype)
        except (ValueError, TypeError):
            msg = ('The output of a metric function can only be a single Tensor. '
                   f'Received: {value}. ')
            if isinstance(value, dict):
                msg += ('To return a dict of values, implement a custom Metric '
                        'subclass.')
            raise RuntimeError(msg)

        with tf.control_dependencies([value]):
            update_total_op = self.total.assign_add(value)

        with tf.control_dependencies([update_total_op]):
            result = self.count.assign_add(1)
        return result

    def result(self):
        return tf.math.divide_no_nan(self.total, self.count)
```

**Comment by geometrikal** [comment]
Update:

It seems this happens the first time a `tf.cast` is encountered in the training step. e.g. changing the relevant lines to 

```
t0 = time.time()
test = tf.cast(np.random.randint(0, 100, (100,100)), dtype=tf.float32)
t1 = time.time()
test = tf.cast(np.random.randint(0, 100, (100, 100)), dtype=tf.float32)
t2 = time.time()
self.loss_3.update_state(1.0)
t3 = time.time()
```

gives a similar result:

```
filters = 8
0.00046s 0.00029s 0.00133s
0.00113s 0.00028s 0.00046s
0.00121s 0.00034s 0.00057s
0.00128s 0.00033s 0.00050s
filters = 32
0.00556s 0.00033s 0.00074s
0.01393s 0.00033s 0.00053s
0.01380s 0.00036s 0.00055s
0.01394s 0.00037s 0.00052s
```

**Comment by geometrikal** [comment]
Closing here and opening as tensorflow issue: https://github.com/tensorflow/tensorflow/issues/53530
